plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -93,14 +117,25 @@ private static SchemaTablePrefix parseTableName(String tableName)
             throws IOException
     {
         if (!isCacheEnabledFor(table.getSchemaTableName())) {
-            return fs.listLocatedStatus(path);
+            return delegate.list(fs, table, path);
+        }
+
+        ValueHolder cachedValueHolder;
+        try {
+            cachedValueHolder = cache.get(path, () -> new ValueHolder(delegate.list(fs, table, path)));
```

Comment:
I find it really interesting the approach of caching the iterator to ensure that in a race context only one and only thread calls for the next file. Great idea!!